### PR TITLE
Add support of Office365 OAuth2.0

### PR DESCRIPTION
**Pull request by mjodhaniSN**
Refernce from :
https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-v2-protocols-oauth-code

Signed-off-by: Maulik Jodhani <maulik.jodhani@softnautics.com>

### CODE REVIEW

File: src/main/java/com/googlesource/gerrit/plugins/oauth/Office365Api.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
+// Copyright (C) 2015 The Android Open Source Project
```
Comment (davido) [inline src/main/java/com/googlesource/gerrit/plugins/oauth/Office365Api.java]: 2018

File: src/main/java/com/googlesource/gerrit/plugins/oauth/Office365OAuthService.java [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+  private static String retrieveHostedDomain(JsonObject jwtToken) {
+    JsonElement hdClaim = jwtToken.get("hd");
+    if (hdClaim != null && !hdClaim.isJsonNull()) {
+      String hd = hdClaim.getAsString();
+      log.debug("OAuth2: hd={}", hd);
+      return hd;
+    }
+    log.debug("OAuth2: JWT doesn't contain hd element");
+    return null;
+  }
```
Comment (davido) [inline src/main/java/com/googlesource/gerrit/plugins/oauth/Office365OAuthService.java]: Does Office 365 really accept hosted domain?

**Comment by mjodhaniSN** [comment]
Hi davido,

Thanks for your review comments.
I have updated patch accordingly and pushed to gerrit-review:
https://gerrit-review.googlesource.com/c/plugins/oauth/+/153930

Thanks,
Maulik

**Comment by davido** [comment]
It was merged and I released a new version that includes this OAuth provider: [1].

* [1] https://github.com/davido/gerrit-oauth-provider/releases/tag/v2.14.6.1

**Pull request closed without merging**